Remove nondeterminism in multiple-definitions test

ricobbe opened the pull request:
Compare all fields in `DllImport` when sorting to avoid nondeterminism in the error for multiple inconsistent definitions of an extern function.  Restore the multiple-definitions test.

Resolves #87084.

rust-highfive commented:
r? @estebank

(rust-highfive has picked a reviewer for you, use r? to override)

ricobbe commented:
r? @petrochenkov

ricobbe commented:
Question for reviewers: in an earlier version of this code, a reviewer asked me not to have `DllImport` implement `PartialOrd` or `Ord`, I think because the ordering relation as implemented then didn't consider all of `DllImport`'s fields, but it's been a while and I don't remember for certain.  The new ordering relation I'm defining here does indeed consider all of the fields, and it's effectively equivalent to the one that we'd get if I just derived `PartialOrd` and `Ord`.  Would you prefer me to just do that instead?

fee1-dead commented:
I don't think it is nondeterministic when you compared the names, so I don't think the issue is that simple.

ricobbe commented:
> I don't think it is nondeterministic when you compared the names, so I don't think the issue is that simple.

The error message is triggered in precisely the case where we have multiple `DllImport`s with the same name (plus some other conditions that aren't important here).  Therefore, to get a unique, well-defined result from sorting the imports (and therefore selecting the second entry with a duplicate name for the error message), I need to compare based on more than just the names.

petrochenkov commented:
My version of the fix, with all unnecessary impl removed, and with explicit sorting removed too - https://github.com/petrochenkov/rust/commit/8b9ce0e7ddf243b8c79022848c1ac85aea9ee900

ricobbe commented:
> My version of the fix, with all unnecessary impls removed, and with explicit sorting removed too - [petrochenkov@8b9ce0e](https://github.com/petrochenkov/rust/commit/8b9ce0e7ddf243b8c79022848c1ac85aea9ee900)

I'm happy to just go with that; what's the workflow to do that?

petrochenkov commented:
@ricobbe
You can cherry-pick that commit and add/update the test.

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -471,16 +472,20 @@ fn collate_raw_dylibs(
-    // when ordering.
-    // FIXME: when we add support for ordinals, figure out if we need to do anything if we
-    // have two DllImport values with the same name but different ordinals.
+    fn compare_imports(x: &DllImport, y: &DllImport) -> Ordering {
```
You can destructure to ensure that you matched all fields. Eg
```suggestion
    fn compare_imports(DllImport { x_name, x_ordinal, x_calling_convention, x_span }: &DllImport, DllImport { y_name, y_ordinal, y_calling_convention, y_span }: &DllImport) -> Ordering {
```
or
```suggestion
    fn compare_imports(x: &DllImport, y: &DllImport) -> Ordering {
        let DllImport { x_name, x_ordinal, x_calling_convention, x_span } = x;
        let DllImport { y_name, y_ordinal, y_calling_convention, y_span } = y;
```

ricobbe replied:
Made moot by petrochenkov's rewrite.

ricobbe commented:
@petrochenkov 

> You can cherry-pick that commit and add/update the test.

Ok, I've tried that, and now the restored (and renamed) multiple-declarations.rs test is failing.  I've run it with `--bless` and updated the expected error message (in the comment in the .rs file) to adjust for the change in wording, but the test still fails: 

```
error: process did not return an error status
status: exit code: 0
```

Everything else looks reasonable; in particular, I'm seeing the expected diagnostics on stderr.

I've tried running the test with the `// check-pass` header command, and also with `// build-pass`, but neither have any effect on the results.  This is not a failure mode I know how to debug, I'm afraid.

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -454,65 +461,46 @@ fn link_rlib<'a, B: ArchiveBuilder<'a>>(
     }
 
-    result
+    if error_detected {
+        Err(ErrorReported)
+    } else {
+        Ok(dylib_table
+            .into_iter()
+            .map(|(name, imports)| {
+                (name, imports.into_iter().map(|(_, import)| import.clone()).collect())
+            })
+            .collect())
+    }
```
```suggestion
    sess.compile_status()?;
    Ok(dylib_table
        .into_iter()
        .map(|(name, imports)| {
            (name, imports.into_iter().map(|(_, import)| import.clone()).collect())
        })
        .collect())
```

petrochenkov commented:
r=me after addressing https://github.com/rust-lang/rust/pull/87092#discussion_r671163114 and squashing commits.

ricobbe commented:
@rustbot -S-waiting-on-author +S-waiting-on-review

ricobbe commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

wesleywiser commented:
@bors r=petrochenkov rollup

bors commented:
:pushpin: Commit ce59f1aac53ba209d0a97af720f596f943fcfbaa has been approved by `petrochenkov`

<!-- @bors r=petrochenkov ce59f1aac53ba209d0a97af720f596f943fcfbaa -->
<!-- homu: {"type":"Approved","sha":"ce59f1aac53ba209d0a97af720f596f943fcfbaa","approver":"petrochenkov"} -->
